rust/src/io/reader.rs

```diff
@@ -208,31 +211,64 @@ impl<'a> FileReader<'a> {
         decoder.decode().await
     }
 
-    /// Read an array of the batch.
-    async fn read_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
-        if let DataType::Dictionary(_, _) = field.data_type() {
-            self.read_dictionary_array(field, batch_id).await
-        } else if field.data_type().is_numeric() {
-            self.read_primitive_array(field, batch_id).await
-        } else if matches!(
-            field.data_type(),
-            DataType::Utf8 | DataType::LargeUtf8 | DataType::Binary | DataType::LargeBinary
-        ) {
-            self.read_binary_array(field, batch_id).await
-        } else {
-            println!("Reading: {}", field);
-            let page_info = self.page_table.get(field.id, batch_id).unwrap();
-            Ok(Arc::new(Int32Array::from(
-                (0..page_info.length as i32).collect::<Vec<_>>(),
-            )))
+    async fn read_struct_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        // TODO: use tokio to make the reads in parallel.
+        let mut sub_arrays = vec![];
+        for child in field.children.as_slice() {
+            // let arr = self.read_array(, batch_id)
+            let arr = self.read_array(&child, batch_id).await?;
+            sub_arrays.push((child.into(), arr));
         }
+
+        Ok(Arc::new(StructArray::from(sub_arrays)))
+    }
+
+    async fn read_list_array(&self, field: &Field, batch_id: i32) -> Result<ArrayRef> {
+        let page_info = self.page_info(field, batch_id)?;
+        let position_arr = self
+            .object_reader
+            .read_primitive_array(&DataType::Int32, page_info.position, page_info.length)
```

Comment:
i guess we could refactor this later into read_list_array and read_large_list_array that calls a parametrized method with idnex type

Reply:
yes, you are right.